### DISCUSSION TITLE
Is it possible to configure runner to output logs in JSON format?

### DISCUSSION THREAD

**Discussion by jacek-jablonski** [actions/runner#3065] [Q&A]
Hi,
Is it possible for the runner to emit logs in JSON format? The text format makes filtering and searching very difficult.

![image](https://github.com/actions/runner/assets/35669512/0e8a3934-0478-4b11-b63b-36b3c45931cf)

**Comment by piotr-tokaj** [comment]
It would be awesome to have this option. Or at least something that makes each log entry print on a single line. Right now JSONs are sometimes output to the logs, but each json entry is on a separate line, making them very hard to parse.

**Comment by gistrec** [comment]
As far as I know, the runner does not currently expose a setting to emit job logs as JSON

For application-level structured logging, the most practical workaround is to emit **JSONL** / newline-delimited JSON: one compact JSON object per line. The GitHub Actions log viewer is line-oriented, so pretty-printed JSON will be split across multiple log lines and becomes harder to parse

For example, avoid this:

```json
{
  "level": "info",
  "message": "something happened"
}
```

Prefer this:

```json
{"level":"info","message":"something happened"}
```

If you already have JSON and want to print it as one line, you can usually compact it before writing to stdout:

```bash
jq -c . result.json
```

For post-processing after the run, GitHub also provides an API endpoint to download job logs:

```http
GET /repos/{owner}/{repo}/actions/jobs/{job_id}/logs
Accept: application/vnd.github+json
```

The downloaded logs are still plain text, not JSON, but this can be useful if you want to parse or ship them elsewhere after the workflow completes

For structured diagnostics inside a workflow, GitHub workflow commands such as `::notice`, `::warning`, and `::error` can create job annotations. Those are not JSON logs either, but they may help if the goal is to make important messages more searchable/visible in the Actions UI
